### PR TITLE
only parse CVideoInfoTag::m_strPictureURL if really needed

### PR DESCRIPTION
**Pull request by Montellese**
## Description
This PR improves the performance of loading the details of video library items from the database.
Right now calling `GetMoviesByWhere()` with e.g. `VideoDbDetailsUniqueID` to also retrieve the unique IDs of the movies is 3 times (or 300%) slower than retrieving all movies without the unique IDs. With the changes in this PR it's only 20% slower.

Apart from refactoring all of `CScraperUrl` to (almost) follow proper object orientation rules the main change involves not automatically parsing the XML contained in `CVideoInfoTag::m_strPictureURL` when loading additional details for video library items from the database. The potential artwork URLs are only really needed when scanning for new content and when opening the "Manage artwork" dialog for a specific video library item. So I've changed the code to only parse the XML contained in `CVideoInfoTag::m_strPictureURL` when it is really needed.

In the end I realized that most of the refactoring of `CScraperUrl` is not really needed so if prefered I can strip this PR down to the minimum which is basically just the last commit (0b6a025b4a6cfda0e809379303de881ace7f7fff).

## Motivation and Context
While looking into loading unique IDs for video library items faster I noticed that in e.g. `CVideoDatabase::GetDetailsForMovie()` we only load the requested details (e.g. unique IDs, ratings, cast, stream details, ...) but we ALWAYS parse the XML contained in `CVideoInfoTag::m_strPictureURL`. There is no way to control it like all the other details.

## How Has This Been Tested?
Manually by making sure that scanning still works and that the potential artwork URLs are still available in the "Manage artwork" dialog for a specific video library item.
The timing measurements have been performed by modifying `CVideoDatabase::GetMoviesByWhere()` to perform the whole movie loading 100 times instead of just once (to get better statistics) and measuring the whole elapsed time. Then I used `VideoLibrary.GetMovies` through JSON-RPC to just retrieve all movies without details and all movies with unique IDs.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Review by DaveTBlake — APPROVED** [review]
Looks fine to me, and happy to keep the refactoring of CScraperURL. 

Only question (of no significance whatsoever) is about naming methods using "Url" rather than "URL". I doubt we have a convention, but since class `CURL` is capitals not "CUrl" I kind of expect to see GetURLs(), AddParsedURL() etc. rather than GetUrls(), AddParsedUrl() etc. But honestly a totally trivial observation

**Comment by Montellese** [comment]
> Only question (of no significance whatsoever) is about naming methods using "Url" rather than "URL". I doubt we have a convention, but since class `CURL` is capitals not "CUrl" I kind of expect to see GetURLs(), AddParsedURL() etc. rather than GetUrls(), AddParsedUrl() etc. But honestly a totally trivial observation

I actually asked myself the same question when I refactored / renamed those methods. But since the class is named `CScraperUrl` and not `CScraperURL` I decided to also go with `Url` in the method names. TBH I find `CURL` a very misleading class name because it looks more like a `#define` to me.

File: xbmc/video/jobs/VideoLibraryRefreshingJob.cpp [hunk 1]
```diff
@@ -332,7 +332,7 @@ bool CVideoLibraryRefreshingJob::Work(CVideoDatabase &db)
     CVideoInfoScanner scanner;
     if (!scanner.RetrieveVideoInfo(items, scanSettings.parent_name,
                                    scraper->Content(), !ignoreNfo,
-                                   scraperUrl.m_url.empty() ? NULL : &scraperUrl,
+                                   scraperUrl.HasUrls() ? &scraperUrl : NULL,
```
Comment (phunkyfish) [inline xbmc/video/jobs/VideoLibraryRefreshingJob.cpp:335]: nullptr

**Review by phunkyfish — CHANGES_REQUESTED** [review]
I had gotten used to URL being used from kodi but I guess it's preference.

Just one minor.

File: xbmc/interfaces/legacy/ListItem.cpp [hunk 2]
```diff
@@ -691,7 +691,7 @@ namespace XBMCAddon
     void ListItem::addAvailableArtwork(std::string url, std::string art_type, std::string preview, std::string referrer, std::string cache, bool post, bool isgz, int season)
     {
       XBMCAddonUtils::GuiLock lock(languageHook, m_offscreen);
-      GetVideoInfoTag()->m_strPictureURL.AddElement(url, art_type, preview, referrer, cache, post, isgz, season);
+      GetVideoInfoTag()->m_strPictureURL.AddParsedUrl(url, art_type, preview, referrer, cache, post, isgz, season);
```
Comment (Rechi) [inline xbmc/interfaces/legacy/ListItem.cpp:694]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/music/Album.cpp [hunk 3]
```diff
@@ -543,12 +543,14 @@ bool CAlbum::Load(const TiXmlElement *album, bool append, bool prioritise)
...
-  if (prioritise && iThumbCount && iThumbCount != thumbURL.m_url.size())
+  if (prioritise && iThumbCount && iThumbCount != thumbURL.GetUrls().size())
   {
-    rotate(thumbURL.m_url.begin(),
-           thumbURL.m_url.begin()+iThumbCount,
-           thumbURL.m_url.end());
-    thumbURL.m_xml = xmlAdd;
+    auto thumbUrls = thumbURL.GetUrls();
+    rotate(thumbUrls.begin(),
+           thumbUrls.begin()+iThumbCount,
```
Comment (Rechi) [inline xbmc/music/Album.cpp:550]: > Conventional operators have to be surrounded by one space on each side.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/music/Artist.cpp [hunk 4]
```diff
@@ -108,12 +108,14 @@ bool CArtist::Load(const TiXmlElement *artist, bool append, bool prioritise)
...
-  if (prioritise && iThumbCount && iThumbCount != thumbURL.m_url.size())
+  if (prioritise && iThumbCount && iThumbCount != thumbURL.GetUrls().size())
   {
-    rotate(thumbURL.m_url.begin(),
-           thumbURL.m_url.begin()+iThumbCount,
-           thumbURL.m_url.end());
-    thumbURL.m_xml = xmlAdd;
+    auto thumbUrls = thumbURL.GetUrls();
+    rotate(thumbUrls.begin(),
+           thumbUrls.begin()+iThumbCount,
```
Comment (Rechi) [inline xbmc/music/Artist.cpp:115]: > Conventional operators have to be surrounded by one space on each side.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 5]
```diff
@@ -1315,7 +1315,7 @@ CMusicInfoScanner::UpdateDatabaseAlbumInfo(CAlbum& album,
         CServiceBroker::GetEventLog().Add(EventPtr(new CMediaLibraryEvent(
           MediaTypeAlbum, album.strPath, 24146,
           StringUtils::Format(g_localizeStrings.Get(24147).c_str(), MediaTypeAlbum, album.strAlbum.c_str()),
-          CScraperUrl::GetThumbURL(album.thumbURL.GetFirstThumb()), CURL::GetRedacted(album.strPath), EventLevel::Warning)));
+          CScraperUrl::GetThumbUrl(album.thumbURL.GetFirstUrlByType()), CURL::GetRedacted(album.strPath), EventLevel::Warning)));
```
Comment (Rechi) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp:1318]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 6]
```diff
@@ -1383,7 +1383,7 @@ CMusicInfoScanner::UpdateDatabaseArtistInfo(CArtist& artist,
         CServiceBroker::GetEventLog().Add(EventPtr(new CMediaLibraryEvent(
           MediaTypeArtist, artist.strPath, 24146,
           StringUtils::Format(g_localizeStrings.Get(24147).c_str(), MediaTypeArtist, artist.strArtist.c_str()),
-          CScraperUrl::GetThumbURL(artist.thumbURL.GetFirstThumb()), CURL::GetRedacted(artist.strPath), EventLevel::Warning)));
+          CScraperUrl::GetThumbUrl(artist.thumbURL.GetFirstUrlByType()), CURL::GetRedacted(artist.strPath), EventLevel::Warning)));
```
Comment (Rechi) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp:1386]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 7]
```diff
@@ -1500,7 +1500,7 @@ CMusicInfoScanner::DownloadAlbumInfo(const CAlbum& album,
       CMusicAlbumInfo albumNfo("nfo",scrUrl);
       ADDON::ScraperPtr nfoReaderScraper = nfoReader.GetScraperInfo();
       CLog::Log(LOGDEBUG,"-- nfo-scraper: %s", nfoReaderScraper->Name().c_str());
-      CLog::Log(LOGDEBUG,"-- nfo url: %s", scrUrl.m_url[0].m_url.c_str());
+      CLog::Log(LOGDEBUG,"-- nfo url: %s", scrUrl.GetFirstThumbUrl());
```
Comment (Rechi) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp:1503]: > Commas have to be followed by one space.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 8]
```diff
@@ -1931,10 +1931,10 @@ bool CMusicInfoScanner::ResolveMusicBrainz(const std::string &strMusicBrainzID,
       return false;
   }
 
-  if (!musicBrainzURL.m_url.empty())
+  if (musicBrainzURL.HasUrls())
   {
     CLog::Log(LOGDEBUG,"-- nfo-scraper: %s",preferredScraper->Name().c_str());
-    CLog::Log(LOGDEBUG,"-- nfo url: %s", musicBrainzURL.m_url[0].m_url.c_str());
+    CLog::Log(LOGDEBUG,"-- nfo url: %s", musicBrainzURL.GetFirstThumbUrl());
```
Comment (Rechi) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp:1937]: > Commas have to be followed by one space.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/video/VideoDatabase.cpp [hunk 9]
```diff
@@ -4764,21 +4756,23 @@ bool CVideoDatabase::GetArtTypes(const MediaType &mediaType, std::vector<std::st
 
 namespace
 {
-std::vector<std::string> GetBasicItemAvailableArtTypes(const CVideoInfoTag& tag)
+std::vector<std::string> GetBasicItemAvailableArtTypes(int mediaId, VIDEODB_CONTENT_TYPE dbType, CVideoDatabase& db)
```
Comment (Rechi) [inline xbmc/video/VideoDatabase.cpp:4759]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/video/VideoInfoDownloader.cpp [hunk 10]
```diff
@@ -78,17 +78,17 @@ void CVideoInfoDownloader::Process()
   else if (m_state == GET_DETAILS)
   {
     if (!GetDetails(m_url, m_movieDetails))
-      CLog::Log(LOGERROR, "%s: Error getting details from %s", __FUNCTION__,m_url.m_url[0].m_url.c_str());
+      CLog::Log(LOGERROR, "%s: Error getting details from %s", __FUNCTION__, m_url.GetFirstThumbUrl());
```
Comment (Rechi) [inline xbmc/video/VideoInfoDownloader.cpp:81]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/video/VideoInfoDownloader.cpp [hunk 11]
```diff
@@ -78,17 +78,17 @@ void CVideoInfoDownloader::Process()
...
   }
   else if (m_state == GET_EPISODE_DETAILS)
   {
     if (!GetEpisodeDetails(m_url, m_movieDetails))
-      CLog::Log(LOGERROR, "%s: Error getting episode details from %s", __FUNCTION__, m_url.m_url[0].m_url.c_str());
+      CLog::Log(LOGERROR, "%s: Error getting episode details from %s", __FUNCTION__, m_url.GetFirstThumbUrl());
```
Comment (Rechi) [inline xbmc/video/VideoInfoDownloader.cpp:86]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/video/VideoInfoDownloader.cpp [hunk 12]
```diff
@@ -78,17 +78,17 @@ void CVideoInfoDownloader::Process()
...
   }
   else if (m_state == GET_EPISODE_LIST)
   {
     if (!GetEpisodeList(m_url, m_episode))
-      CLog::Log(LOGERROR, "%s: Error getting episode list from %s", __FUNCTION__, m_url.m_url[0].m_url.c_str());
+      CLog::Log(LOGERROR, "%s: Error getting episode list from %s", __FUNCTION__, m_url.GetFirstThumbUrl());
```
Comment (Rechi) [inline xbmc/video/VideoInfoDownloader.cpp:91]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/video/VideoInfoScanner.cpp [hunk 13]
```diff
@@ -1729,7 +1729,7 @@ namespace VIDEO
           matches.push_back(*guide);
           continue;
         }
-        if (!guide->cScraperUrl.strTitle.empty() && StringUtils::EqualsNoCase(guide->cScraperUrl.strTitle, file->strTitle))
+        if (!guide->cScraperUrl.GetTitle().empty() && StringUtils::EqualsNoCase(guide->cScraperUrl.GetTitle(), file->strTitle))
```
Comment (Rechi) [inline xbmc/video/VideoInfoScanner.cpp:1732]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/video/VideoInfoTag.cpp [hunk 14]
```diff
@@ -985,12 +987,14 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
...
-  if (prioritise && iThumbCount && iThumbCount != m_strPictureURL.m_url.size())
+  if (prioritise && iThumbCount && iThumbCount != m_strPictureURL.GetUrls().size())
   {
-    rotate(m_strPictureURL.m_url.begin(),
-           m_strPictureURL.m_url.begin()+iThumbCount,
-           m_strPictureURL.m_url.end());
-    m_strPictureURL.m_xml = xmlAdd;
+    auto thumbUrls = m_strPictureURL.GetUrls();
+    rotate(thumbUrls.begin(),
+           thumbUrls.begin()+iThumbCount,
```
Comment (Rechi) [inline xbmc/video/VideoInfoTag.cpp:994]: > Conventional operators have to be surrounded by one space on each side.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/video/jobs/VideoLibraryRefreshingJob.cpp [hunk 15]
```diff
@@ -212,7 +212,7 @@ bool CVideoLibraryRefreshingJob::Work(CVideoDatabase &db)
               return false;
           }
 
-          CLog::Log(LOGDEBUG, "CVideoLibraryRefreshingJob: user selected item '%s' with URL '%s'", scraperUrl.strTitle.c_str(), scraperUrl.m_url.at(0).m_url.c_str());
+          CLog::Log(LOGDEBUG, "CVideoLibraryRefreshingJob: user selected item '%s' with URL '%s'", scraperUrl.GetTitle().c_str(), scraperUrl.GetFirstThumbUrl());
```
Comment (Rechi) [inline xbmc/video/jobs/VideoLibraryRefreshingJob.cpp:215]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

File: xbmc/video/jobs/VideoLibraryRefreshingJob.cpp (hunk 1 above)
Comment (Montellese) [reply xbmc/video/jobs/VideoLibraryRefreshingJob.cpp]: Fixed

File: xbmc/interfaces/legacy/ListItem.cpp (hunk 2 above)
Comment (Montellese) [reply xbmc/interfaces/legacy/ListItem.cpp:694]: Fixed

File: xbmc/video/VideoInfoScanner.cpp (hunk 13 above)
Comment (Montellese) [reply xbmc/video/VideoInfoScanner.cpp:1732]: Fixed

File: xbmc/video/jobs/VideoLibraryRefreshingJob.cpp (hunk 15 above)
Comment (Montellese) [reply xbmc/video/jobs/VideoLibraryRefreshingJob.cpp:215]: Fixed

File: xbmc/music/Album.cpp (hunk 3 above)
Comment (Montellese) [reply xbmc/music/Album.cpp:550]: Fixed

File: xbmc/music/Artist.cpp (hunk 4 above)
Comment (Montellese) [reply xbmc/music/Artist.cpp:115]: Fixed

File: xbmc/music/infoscanner/MusicInfoScanner.cpp (hunk 5 above)
Comment (Montellese) [reply xbmc/music/infoscanner/MusicInfoScanner.cpp:1318]: Fixed

File: xbmc/music/infoscanner/MusicInfoScanner.cpp (hunk 6 above)
Comment (Montellese) [reply xbmc/music/infoscanner/MusicInfoScanner.cpp:1386]: Fixed

File: xbmc/music/infoscanner/MusicInfoScanner.cpp (hunk 7 above)
Comment (Montellese) [reply xbmc/music/infoscanner/MusicInfoScanner.cpp:1503]: Fixed

File: xbmc/music/infoscanner/MusicInfoScanner.cpp (hunk 8 above)
Comment (Montellese) [reply xbmc/music/infoscanner/MusicInfoScanner.cpp:1937]: Fixed

File: xbmc/video/VideoInfoDownloader.cpp (hunk 10 above)
Comment (Montellese) [reply xbmc/video/VideoInfoDownloader.cpp:81]: Fixed

File: xbmc/video/VideoInfoDownloader.cpp (hunk 11 above)
Comment (Montellese) [reply xbmc/video/VideoInfoDownloader.cpp:86]: Fixed

File: xbmc/video/VideoInfoDownloader.cpp (hunk 12 above)
Comment (Montellese) [reply xbmc/video/VideoInfoDownloader.cpp:91]: Fixed

File: xbmc/video/VideoInfoTag.cpp (hunk 14 above)
Comment (Montellese) [reply xbmc/video/VideoInfoTag.cpp:994]: Fixed

File: xbmc/video/VideoDatabase.cpp (hunk 9 above)
Comment (Montellese) [reply xbmc/video/VideoDatabase.cpp:4759]: Fixed

**Review by phunkyfish — APPROVED** [review]
LGTM

**Comment by Montellese** [comment]
I have addressed all comments. I want to test one more case and then this should be good to go.
jenkins build this please

**Comment by Montellese** [comment]
OK I checked the case I mentioned and it looks good.

@notspiff You are probably the only one who was involved back when this was introduced. Any inputs or red flags you remember or does this seem ok for you?

**Comment by notspiff** [comment]
No concerns. In fact this is how it was back when i wrote the code.
Countless refactorings lost this optimization i guess..

tor. 14. mai 2020, 17:43 skrev Sascha Montellese <notifications@github.com>:

**Comment by Montellese** [comment]
@notspiff thanks for the feedback.

**Pull request merged**